refactor MetaData.Builder to update aliasAndIndexLookup

talevy opened the pull request:
It would be nice to be able to look up other indices' aliases
while validating new aliases being introduced. For example,
If we would like to prohibit certain aliases and allow others,
depending on their respective definitions.

The problem with moving the construction of this lookup data-structure
to be iterative is that we do not have a clear view of *all* other possible
duplicate aliases. We would raise an error with the first duplicate we see.

talevy commented:
@martijnvg after realizing what it takes to make this change work, I'm leaning towards your suggestion that was made offline about having more validation in `Metadata.Builder#build` instead of the other way around. The need for this data-structure being updated as one builds `MetaData` would not be necessary if more validation was just added to the `build` method. I like the simplicity of this, but I am not sure how costly it is to add some extra loops here for every build. what do you think?

bleskes commented:
@talevy @martijnvg I understand your sentiment but there are also things like https://github.com/elastic/elasticsearch/issues/27689 we should consider. We allow a list of atomic alias updates that are applied one by one. I think we need to truly execute them one by one while tracking the current state and throwing errors as soon as we get to a bad place. For example, what happens if people send two commands to update the write alias for the same alias? what will the second one use as a basis?

>  I like the simplicity of this, but I am not sure how costly it is to add some extra loops here for every build. what do you think?

It should be sane (i.e. not a n^2 operation) but also doesn't need to be super high performance.

> If we feel like more validation can be added to the build method

As discussed validation in the build method is required regardless. We don't always construct cluster states via the formal API (i.e., during recovery we don't).

talevy commented:
@bleskes I see, thanks. I will push forward with this. I believe I have fixed all the test failures after my last commit.

talevy commented on server/src/main/java/org/elasticsearch/cluster/metadata/AliasAndIndexLookupBuilder.java:
```diff
@@ -0,0 +1,145 @@
+                    return alias;
+                } else {
+                    String indexName = ((AliasOrIndex.Index) alias).getIndex().getIndex().getName();
+                    throw new IllegalStateException("index and alias names need to be unique, but the following duplicate was found ["
```
I understand I ported this over from `MetaData.Builder#build` where `IllegalStateException` makes more sense. I believe this should be `IllegalAliasNameException`, will update

talevy commented:
@elasticmachine please test this

talevy commented:
So, I've moved along with the `is_write_index` feature that brought this refactor to my attention in the first place. I've implemented that with the existing "rebuild upon request" strategy [1]. If we feel like rebuilding each time is not a big deal, then I can go ahead and close this PR.
Let me know!

[1] https://github.com/talevy/elasticsearch/blob/ece9f506818e1bc6758ac3174da7a39946cebdd6/server/src/main/java/org/elasticsearch/cluster/metadata/AliasValidator.java#L185

elasticmachine commented:
Pinging @elastic/es-distributed

talevy commented:
After discussion with the team, I am closing this. It is not necessary for me anymore
